### PR TITLE
Initial 3LC integration

### PR DESCRIPTION
**Pull request by fredmell**
This PR adds the initial 3LC integration into YOLOv8. Currently only object detection is supported.

The initial integration supports most of the functionality available in the YOLOv5 integration, except for collecting loss during metrics collection. A key difference between  the two integrations is that YOLOv5 accepts 3LC options through environment variables - here they are provided through an instance of `Settings`.

For now one has to clone this fork and use a local install of the package (`pip install -e .`).

To train, the following code can be used:
```python
from ultralytics.utils.tlc.detect.model import TLCYOLO  # noqa: E402
from ultralytics.utils.tlc.detect.settings import Settings

model = TLCYOLO("yolov8n.pt")  # initialize

# Set 3LC specific settings
settings = Settings(
    image_embeddings_dim=2,
    collection_epoch_start=0,
    collection_epoch_interval=2,
    conf_thres=0.2,
)

# Run training
results = model.train(
    data="coco128.yaml",
    device=0,
    epochs=5,
    batch=32,
    imgsz=320,
    workers=0,
    settings=settings,
)
```

To only collect metrics, the following can be used:
```python
from ultralytics.utils.tlc.detect.settings import Settings
from ultralytics.utils.tlc.detect.model import TLCYOLO

model = TLCYOLO("yolov8l.pt")

# Set 3LC specific settings
settings = Settings(
    image_embeddings_dim=2,
    conf_thres=0.2,
)

for split in ("train", "val"):
    results = model.val(data="coco128.yaml", split=split, batch=32, imgsz=320, device=0, workers=0, settings=settings)
```
Note that for validation only, on multiple calls to `model.val()`, the same run will be used across calls. Each call should correspond to the name of a split in the dataset being used.

### CODE REVIEW

File: ultralytics/utils/tlc/detect/nn.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+class TLCDetectionModel(DetectionModel):
+    activations = None
+    """ YOLO (You Only Look Once) object detection model with 3LC integration. """
+
+    def _predict_once(self, x, profile=False, visualize=False, embed=None):
+        """
+        Perform a forward pass through the network.
+
+        Args:
```
Comment (gudbrandtandberg) [inline ultralytics/utils/tlc/detect/nn.py:17]: Does not quite match the format of docstrings in other files

File: ultralytics/utils/tlc/detect/utils.py [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+
+
+def get_tlc_table_from_url(table_url: tlc.Url, split: str) -> tlc.Table:
+    """ Get a 3LC table from a URL.
+
+    :param table_url: The Url of the table.
+    :param split: The split the table corresponds to.
+    :returns: The 3LC table.
+    :raises: ValueError if the table does not exist.
+    :raises: ValueError if the table is not compatible with YOLOv5.
```
Comment (fredmell) [inline ultralytics/utils/tlc/detect/utils.py:248]: This should be v8

File: ultralytics/utils/tlc/detect/model.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+from ultralytics.utils.tlc.detect.nn import TLCDetectionModel
+from ultralytics.utils.tlc.detect.trainer import TLCDetectionTrainer
+from ultralytics.utils.tlc.detect.utils import tlc_check_dataset
+from ultralytics.utils.tlc.detect.validator import TLCDetectionValidator
+
+
+def check_det_dataset(data: str):
+    """Check if the dataset is compatible with the 3LC."""
+    tables = tlc_check_dataset(data)
+    names = tables["train"].get_value_map_for_column(tlc.BOUNDING_BOXES)
```
[the remark this answers is not in the document]
Comment (fredmell) [reply ultralytics/utils/tlc/detect/model.py]: Good point, fixed!
